server/src/main/java/org/apache/druid/segment/realtime/appenderator/UnifiedIndexerAppenderatorsManager.java

```diff
@@ -183,22 +192,48 @@ public Appenderator createOfflineAppenderatorForTask(
       IndexMerger indexMerger
   )
   {
-    Appenderator appenderator = Appenderators.createOffline(
-        schema,
-        rewriteAppenderatorConfigMemoryLimits(config),
-        metrics,
-        dataSegmentPusher,
-        objectMapper,
-        indexIO,
-        wrapIndexMerger(indexMerger)
-    );
-    return appenderator;
+    synchronized (this) {
+      DatasourceBundle datasourceBundle = datasourceBundles.computeIfAbsent(
+          schema.getDataSource(),
+          (datasource) -> {
+            return new DatasourceBundle(datasource);
+          }
+      );
+
+      Appenderator appenderator = Appenderators.createOffline(
+          schema,
+          rewriteAppenderatorConfigMemoryLimits(config),
+          metrics,
+          dataSegmentPusher,
+          objectMapper,
+          indexIO,
+          wrapIndexMerger(indexMerger)
+      );
+      datasourceBundle.addAppenderator(taskId, appenderator);
+      return appenderator;
+    }
   }
 
   @Override
-  public void removeAppenderatorForTask(String taskId)
+  public void removeAppenderatorsForTask(
+      String taskId,
+      String dataSource
+  )
   {
-    // nothing to remove presently
+    synchronized (this) {
+      DatasourceBundle datasourceBundle = datasourceBundles.get(dataSource);
+      if (datasourceBundle == null) {
+        LOG.error("Could not find datasource bundle for [%s], task [%s]", dataSource, taskId);
```

Comment:
Wondering how fatal this case is. I think, if it's very fatal, it should be propagated to the task so that it will be logged in the task log file in the future. Otherwise, perhaps the warning level would be enough.

Reply:
I think this is more of an error at the level of the task manager and it would only be called after a task has finished (so I don't think it could be propagated to tasks), it's not fatal enough to stop altogether but I think it indicates bugs so I wanted to be ERROR instead of WARN.

Reply:
Hmm, I see. Then perhaps it would be better to emit an alert using `EmittingLogger` so that operators could easily notice this error. I'm worried about that this could mean resource leaks but ppl would never read logs until something happens. What do you think?

Reply:
Hm, I think it's fine as ERROR for now since the operator can't really take any action to remedy the situation, and I don't think the error is severe enough to warrant alerting. We could revisit this area in the future if needed.

Reply:
👍